python/lsst/daf/butler/remote_butler/registry/_query_dimension_records.py

```diff
@@ -0,0 +1,137 @@
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+import dataclasses
+from collections.abc import Callable, Iterable, Iterator
+from contextlib import AbstractContextManager, contextmanager
+from typing import Any, TypeAlias
+
+from ...dimensions import DataId, DataIdValue, DimensionElement, DimensionRecord
+from ...queries import DimensionRecordQueryResults, Query
+from ...registry.queries import DimensionRecordQueryResults as LegacyDimensionRecordQueryResults
+
+QueryFactory: TypeAlias = Callable[[], AbstractContextManager[Query]]
+"""Function signature matching the interface of ``Butler._query``.  Returns a
+context manager that can be used to obtain a `Query` instance.
+"""
+
+
+@dataclasses.dataclass(frozen=True)
+class DimensionRecordsQueryArguments:
+    """Simplified version of the function arguments passed to
+    ``Registry.queryDimensionRecords``.
+    """
+
+    element: DimensionElement
+    dataId: DataId | None
+    dataset_types: list[str]
+    collections: list[str] | None
+    where: str
+    bind: dict[str, Any] | None
+    kwargs: dict[str, DataIdValue]
+
+
+class QueryDriverDimensionRecordQueryResults(LegacyDimensionRecordQueryResults):
+    """Implementation of the legacy ``DimensionRecordQueryResults`` interface
+    using the new query system.
+
+    Parameters
+    ----------
+    query_factory : `QueryFactory`
+        Function that can be called to access the new query system.
+    args : `DimensionRecordsQueryArguments`
+        User-facing arguments forwarded from
+        ``registry.queryDimensionRecords``.
+    """
+
+    def __init__(
+        self,
+        query_factory: QueryFactory,
+        args: DimensionRecordsQueryArguments,
+    ) -> None:
+        self._query_factory = query_factory
+        self._args = args
+        self._limit: int | None = None
+        self._order_by: list[str] = []
+
+    @property
+    def element(self) -> DimensionElement:
+        return self._args.element
+
+    def __iter__(self) -> Iterator[DimensionRecord]:
+        with self._build_query() as result:
+            # We have to eagerly fetch the results to prevent
+            # leaking the resources associated with QueryDriver.
+            records = list(result)
+        yield from records
```

Comment:
I suspect (but do not know) that `return iter(records)` is a little more efficient in cases like this.